### PR TITLE
object: add the external rgw server check properly

### PR DESCRIPTION
**Pull request by thotz**
**Description of your changes:**
In object package code, external rgw server checks whether ceph cluster
is external instead of rgw. Correcting such scenarios in the code base.

Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/object/controller.go [hunk 1]
```diff
@@ -275,15 +275,15 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 
 		// Cancel the context to stop monitoring the health of the object store
 		r.stopMonitoring(cephObjectStore)
-
-		cfg := clusterConfig{
-			context:     r.context,
-			store:       cephObjectStore,
-			clusterSpec: r.clusterSpec,
-			clusterInfo: r.clusterInfo,
+		if !cephObjectStore.Spec.IsExternal() {
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:278]: Inside the `deleteStore()` method it already checks for external cluster, seems like we don't need to duplicate the check here.

File: pkg/operator/ceph/object/user/controller.go [hunk 2]
```diff
@@ -571,10 +571,8 @@ func (r *ReconcileObjectStoreUser) validateUser(u *cephv1.CephObjectStoreUser) e
 	if u.Namespace == "" {
 		return errors.New("missing namespace")
 	}
-	if !r.cephClusterSpec.External.Enable {
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go:574]: This should really be checked for both external and local clusters? Seems like external cluster wouldn't expect u.Spec.Store to be set.

**Review by travisn — CHANGES_REQUESTED** [review]
Is there a scenario where we expect cluster CR IsExternal() to be different from the object store IsExternal()? They are always the same scenario, right? This change is just for consistency?

**Comment by thotz** [comment]
> Is there a scenario where we expect cluster CR IsExternal() to be different from the object store IsExternal()? They are always the same scenario, right? This change is just for consistency?

One of ideas behind making it consistent, users will be able to set up an external rgw server without configuring external ceph cluster. Currently, it won't work since ceph version requires `clusterspec`

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (thotz) [reply pkg/operator/ceph/object/user/controller.go:574]: Sorry I don't understand that, is there any specific reason for that? For the external rgw server, we do have CephObjectStore() CR right?

Comment (BlaineEXE) [reply pkg/operator/ceph/object/user/controller.go:574]: I guess the question in my mind is whether we should be ignoring the CephCluster's external configuration from the reconcile. From what I gather, this makes it so we only care about the external status of the CephObjectStore. I'm still considering the overall ramifications this might have, positively or negativeluy.

Comment (BlaineEXE) [reply pkg/operator/ceph/object/user/controller.go:574]: I am having concerns about the number of external permutations that exist for CephCluster and CephObjectStore.

With code released to date, I believe we have 3 operating permutations:
1. Local CephCluster with Local CephObjectStore
2. External-mode CephCluster with Local CephObjectStore (mons and osds are remote, but we run the RGW inside k8s) (https://rook.github.io/docs/rook/v1.9/CRDs/Cluster/external-cluster/#cephcluster-example-management)
3. External-mode CephCluster with external-mode CephObjectStore

I think all 3 of these operating permutations make sense. But I think that this PR may be opening up a 4th operating permutation, and I'm not sure this 4th operating mode makes sense as a supportable configuration:
4. Local CephCluster with external-mode CephObjectStore

Comment (parth-gr) [reply pkg/operator/ceph/object/user/controller.go:574]: > I am having concerns about the number of external permutations that exist for CephCluster and CephObjectStore.
> 
> With code released to date, I believe we have 3 operating permutations:
> 
> 1. Local CephCluster with Local CephObjectStore
> 2. External-mode CephCluster with Local CephObjectStore (mons and osds are remote, but we run the RGW inside k8s) ([doc](https://rook.github.io/docs/rook/v1.9/CRDs/Cluster/external-cluster/#cephcluster-example-management))
> 3. External-mode CephCluster with external-mode CephObjectStore
> 
> I think all 3 of these operating permutations make sense. But I think that this PR may be opening up a 4th operating permutation, and I'm not sure this 4th operating mode makes sense as a supportable configuration: 4. Local CephCluster with external-mode CephObjectStore

This comment really make sense, what I was talking about

File: pkg/operator/ceph/object/objectstore.go [hunk 3]
```diff
@@ -875,11 +875,9 @@ func getDashboardUser(context *Context) (ObjectUser, error) {
 		SystemUser:  true,
 	}
 
-	if !context.CephClusterSpec.External.Enable {
```
Comment (parth-gr) [inline pkg/operator/ceph/object/objectstore.go:878]: Maybe these checks were added so we can prevent reconciling External-cluster against rgw reconciliation process as it doesn't work probably for now.

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/user/controller.go:574]: I believe external mode was originally created with case 2 working, and I believe it will still work today, though we certainly don't have any CI tests that validate this.

Comment (thotz) [reply pkg/operator/ceph/object/user/controller.go:574]: @travisn @BlaineEXE @parth-gr Actually, this check has nothing to do with the different modes mentioned, here it validates `cephobjectstoreuser`  CR. There is nothing special for external or local rgw while creating/provisioning `cephobjectstoreUser` both does it with help of adminops API. So IMO to include a check for both.

@BlaineEXE: I not trying to create a different mode not in this PR for sure. And I guess mode 2 should work today not sure whether it is desired or not. What in my mind was to remove the dependency for `Cephcluster` CR for the external rgw server, again not with this PR. I felt External `CephObjectStore CR` can work without knowing ceph details or running ceph commands because Rook considers rgw as an endpoint for the external use case.

**Comment by travisn** [comment]
> > Is there a scenario where we expect cluster CR IsExternal() to be different from the object store IsExternal()? They are always the same scenario, right? This change is just for consistency?
> 
> One of ideas behind making it consistent, users will be able to set up an external rgw server without configuring external ceph cluster. Currently, it won't work since ceph version requires `clusterspec`. IMO the ceph version check may not be needed for the external rgw server, but I still need to investigate because Rook communicates with rgw via rest apis or s3 apis not via ceph commands

It's an interesting question, whether the CephCluster CR must be created before creating the external object store CR. It does seem independent as long as Rook is only communicates with rgw via rest apis or s3 as you said. If we don't require the CephCluster CR, it does open the door for another which is to connect to the external rgw even while you have a local cephcluster CR, but I don't see why we would need to restrict that either. In fact, to validate that this PR is working, seems like we should not create the CephCluster CR and verify the external RGW is working, right?

**Comment by BlaineEXE** [comment]
I think we should rename the commit and title of this PR as well. "add the external rgw server check properly" makes it seem as if this PR is fixing a bug, but it is actually changing behavior of the CephObjectStore's external mode. This is a substantial thing to note for users.

IMO, it's probably good to mention this in the pending release notes, and I don't think we need to backport this to 1.9. It will give some extra time for us to find any little bugs as well if we wait to release it until 1.10.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
In order to fully convert to `CephCluster` optional behavior, I think we may need to change how `IsReadyToReconcile()` is called in the controller. I think the right thing to do would be to only call this if the CephObjectStore is not in external mode.

Additionally, in order to make sure we aren't making the rest of the CephObjectStore code reliant on the CephCluster, we should remove `clusterSpec` from `clusterConfig`. That will help us better understand what functions need the CephCluster spec and make sure we aren't dependent on it for external configurations. For any functions/methods that need the CephCluster spec, we should pass it to those functions via a parameter.

File: pkg/operator/ceph/object/controller.go [hunk 4]
```diff
@@ -275,15 +275,13 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
...
 		r.stopMonitoring(cephObjectStore)
-
 		cfg := clusterConfig{
 			context:     r.context,
 			store:       cephObjectStore,
 			clusterSpec: r.clusterSpec,
 			clusterInfo: r.clusterInfo,
 		}
 		cfg.deleteStore()
-
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/controller.go:286]: Why remove these spaces? I think they separate the pre-existing content well.

**Review by BlaineEXE — COMMENTED** [review]
Overall, I think this looks good. I'd like to get an answer to Parth's question here: https://github.com/rook/rook/pull/10425/files#r921357696

And it should be a quick fix to make sure PRs aren't changing unrelated files (removing spaces in ceph/object/controller).

File: pkg/operator/ceph/object/objectstore.go (hunk 3 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/objectstore.go:878]: I think this is a good question to investigate. This is related to dashboard. Do we know if dashboard will work for external clusters?

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (yuvalif) [reply pkg/operator/ceph/object/user/controller.go:574]: @thotz this makes sense. but maybe to avoid "option 4" (at least in this PR) we should also make sure that the cluster is external as well?

Comment (thotz) [reply pkg/operator/ceph/object/user/controller.go:574]: @yuvalif : Actually, there is a bug in the current implementation 
Even if we check for the cluster is external, we cannot make sure the rgw server belongs to that external ceph cluster.
Say I have created `cephcluster` CR with external ceph cluster A and with current checks I can still create `cephObjectStore` CR for rgw server which belongs to another external ceph cluster B.
The plan was to fix it with help https://github.com/ceph/go-ceph/pull/616

File: pkg/operator/ceph/object/objectstore.go (hunk 3 above)
Comment (thotz) [reply pkg/operator/ceph/object/objectstore.go:878]: getDashboardUser() is called from enableDashboard() which is called from createOrUpdateStore(). The createOrUpdateStore() called only for internal cluster, so we are not enabling dashboard for external rgw servers

File: pkg/operator/ceph/object/controller.go [hunk 5]
```diff
@@ -282,8 +282,8 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 			clusterSpec: r.clusterSpec,
 			clusterInfo: r.clusterInfo,
 		}
-		cfg.deleteStore()
 
+		cfg.deleteStore()
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/controller.go:286]: Undo this change please.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

**Review by travisn — APPROVED** [review]

**Review by BlaineEXE — APPROVED** [review]
